python/mxnet/gluon/model_zoo/text/lm.py

```diff
@@ -15,110 +15,287 @@
+            self.embedding = self._get_embedding()
+            self.encoder = self._get_encoder()
+            self.decoder = self._get_decoder()
 
     def _get_embedding(self):
         embedding = nn.HybridSequential()
         with embedding.name_scope():
-            embedding.add(nn.Embedding(len(self._src_vocab), self._embed_dim,
+            embedding.add(nn.Embedding(self._vocab_size, self._embed_size,
                                        weight_initializer=init.Uniform(0.1)))
             if self._dropout:
                 embedding.add(nn.Dropout(self._dropout))
         return embedding
 
     def _get_encoder(self):
-        encoder = ExtendedSequential()
-        with encoder.name_scope():
-            for l in range(self._num_layers):
-                encoder.add(get_rnn_layer(self._mode, 1, self._embed_dim if l == 0 else
-                                          self._hidden_dim, self._hidden_dim if
-                                          l != self._num_layers - 1 or not self._tie_weights
-                                          else self._embed_dim, 0, 0))
-
-        return encoder
+        return get_rnn_layer(self._mode, self._num_layers, self._embed_size,
+                             self._hidden_size, self._dropout, 0)
 
     def _get_decoder(self):
-        vocab_size = len(self._tgt_vocab)
         if self._tie_weights:
-            output = nn.Dense(vocab_size, flatten=False, params=self.embedding[0].params)
+            output = nn.Dense(self._vocab_size, flatten=False, params=self.embedding[0].params)
         else:
-            output = nn.Dense(vocab_size, flatten=False)
+            output = nn.Dense(self._vocab_size, flatten=False)
         return output
 
     def begin_state(self, *args, **kwargs):
-        return self.encoder[0].begin_state(*args, **kwargs)
+        return self.encoder.begin_state(*args, **kwargs)
+
+    def forward(self, inputs, begin_state=None): # pylint: disable=arguments-differ
+        embedded_inputs = self.embedding(inputs)
+        if not begin_state:
+            begin_state = self.begin_state(batch_size=inputs.shape[1])
+        encoded, state = self.encoder(embedded_inputs, begin_state)
+        out = self.decoder(encoded)
+        return out, state
+
+
+def _load_vocab(dataset_name, vocab, root):
+    if dataset_name:
+        if vocab is not None:
+            warnings.warn('Both dataset_name and vocab are specified. Loading vocab for dataset. '
+                          'vocab will be ignored.')
+        vocab = _load_pretrained_vocab(dataset_name, root)
+    else:
+        assert vocab is not None, "Must specify vocab if not loading from predefined datasets."
+    return vocab
+
+
+def _load_pretrained_params(net, model_name, root, ctx):
+    model_file = get_model_file(model_name, root=root)
+    net.load_params(model_file, ctx=ctx)
+
+
+def _get_rnn_model(model_cls, model_name, dataset_name, vocab, pretrained, ctx, root, **kwargs):
+    if 'vocab_size' not in kwargs:
```

Comment:
Seems to be a typo. Should be 'vocab'?

Reply:
This is not advertised in the api doc. It's left for convenience.

Reply:
model_name alone can't decide the vocabulary size.

Reply:
I'm removing this func to save the trouble of debate